cmds/exp/nvme_unlock/nvme_unlock.go

```diff
@@ -0,0 +1,172 @@
+// unlocked partitions.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"regexp"
+	"syscall"
+	"unsafe"
+
+	"github.com/u-root/u-root/pkg/hsskey"
+	"github.com/u-root/u-root/pkg/mount/block"
+)
+
+const (
+	opalLockUnlockIoctl = 1092120797
+)
+
+type opalKey struct {
+	lr     byte
+	keyLen byte
+	align  [6]byte
+	key    [256]byte
+}
+
+type opalSessionInfo struct {
+	sum     uint32
+	who     uint32
+	opalKey opalKey
+}
+
+type opalLockUnlock struct {
+	session opalSessionInfo
+	lState  uint32
+	align   [4]byte
+}
+
+var (
+	disk               = flag.String("disk", "/dev/nvme0n1", "The disk to be unlocked")
+	verbose            = flag.Bool("d", false, "print debug output")
+	verboseNoSanitize  = flag.Bool("dangerously-disable-sanitize", false, "Print sensitive information - this should only be used for testing!")
+	noRereadPartitions = flag.Bool("no-reread-partitions", false, "Only attempt to unlock the disk, don't re-read the partition table.")
+	lock               = flag.Bool("lock", false, "Lock instead of unlocking")
+)
+
+func verboseLog(msg string) {
+	if *verbose {
+		log.Print(msg)
+	}
+}
+
+func getSysfsInfo(index string, field string) ([]byte, error) {
+	data, err := os.ReadFile(fmt.Sprintf("/sys/class/nvme/nvme%s/%s", index, field))
+	if err != nil {
+		return nil, err
+	}
+	for i, b := range data {
+		if b == ' ' || b == '\n' {
+			return data[0:i], nil
+		}
+	}
+	return data, nil
+}
+
+func main() {
+	flag.Parse()
+
+	commandName := "unlock"
+	if *lock {
+		commandName = "lock"
+	}
+
+	diskIDRegexp := regexp.MustCompile(`/dev/nvme(\d+)n.*`)
+	diskIDMatches := diskIDRegexp.FindStringSubmatch(*disk)
+	diskID := diskIDMatches[1]
+
+	serial, err := getSysfsInfo(diskID, "serial")
+	if err != nil {
+		log.Printf("error getting sysfs info for serial: %v", err)
+	}
+	model, err := getSysfsInfo(diskID, "model")
+	if err != nil {
+		log.Printf("error getting sysfs info for model: %v", err)
+	}
+
+	verboseLog(fmt.Sprintf("Serial %s", serial))
+	verboseLog(fmt.Sprintf("Model %s", model))
+
+	diskFd, err := os.Open(*disk)
+	if err != nil {
+		log.Fatalf("error opening disk: %v", err)
+	}
+
+	hssList, err := hsskey.GetAllHss(*verbose, *verboseNoSanitize)
+	if err != nil {
+		log.Fatalf("error getting HSS: %v", err)
+	}
+
+	if len(hssList) == 0 {
+		log.Fatalf("no HSS found - can't unlock disk.")
+	}
+
+	verboseLog(fmt.Sprintf("Found %d Host Secret Seeds.", len(hssList)))
+
+	succeeded := false
+	for i, hss := range hssList {
+		password, err := hsskey.GenPassword(hss, serial, model)
+		if err != nil {
+			log.Printf("Couldn't generate password with HSS %d: %v", i, err)
+			continue
+		}
+
+		var state uint32 = 0x02 // OPAL_RW
+		if *lock {
+			state = 0x04 // OPAL_LK
+		}
+		arg := opalLockUnlock{
+			session: opalSessionInfo{
+				sum: 0,
+				who: 0,
+				opalKey: opalKey{
+					keyLen: 32,
+				},
+			},
+			lState: state,
+		}
+		for i := 0; i < 32; i++ {
+			arg.session.opalKey.key[i] = password[i]
+		}
```

Comment:
can you use `copy(arg.session.opalKey.key[0:31], password)` ?

Reply:
Fixed.